src/libcore/marker.rs

```diff
@@ -672,6 +673,49 @@ mod impls {
     unsafe impl<T: Send + ?Sized> Send for &mut T {}
 }
 
+/// Compiler-internal trait used to indicate the type of enum discriminants.
+///
+/// This trait is automatically implemented for every type and does not add any
+/// guarantees to [`mem::Discriminant`]. It is **undefined behavior** to transmute
+/// between `DiscriminantKind::Discriminant` and `mem::Discriminant`.
+///
+/// [`mem::Discriminant`]: https://doc.rust-lang.org/stable/core/mem/struct.Discriminant.html
+#[unstable(
+    feature = "discriminant_kind",
+    issue = "none",
+    reason = "this trait is unlikely to ever be stabilized, use `mem::discriminant` instead"
+)]
+#[cfg_attr(not(bootstrap), lang = "discriminant_kind")]
+pub trait DiscriminantKind {
+    /// The type of the dicriminant, which must satisfy the trait
+    /// bounds required by `mem::Discriminant`.
+    type Discriminant: Clone + Copy + Debug + Eq + PartialEq + Hash + Send + Sync + Unpin;
+}
+
+// This default impl is always overwritten by the compiler.
+//
+// We have to use specialization here, as we want to allow
+// `<T as DiscriminantKind>::Discriminant` without adding a trait bound
+// to `T`. We can't just add a builtin implementation for each type on the
+// fly, as the concrete associated type is not always known during typeck.
+//
+// This also prevents any external crates from manually implementing this trait.
+#[unstable(
+    feature = "discriminant_kind",
+    issue = "none",
+    reason = "this trait is unlikely to ever be stabilized, use `mem::discriminant` instead"
+)]
+#[doc(hidden)]
+impl<T: ?Sized> DiscriminantKind for T {
+    #[cfg(bootstrap)]
+    type Discriminant = u64;
+
+    // We use `!` as a safety net, as compiler bugs can now be detected
+    // by checking for an uninhabited type during codegen/const-eval.
+    #[cfg(not(bootstrap))]
+    default type Discriminant = !;
```

Comment:
This is not a good idea AFAIK, it'd be too easy to end up with UB from `!` showing up in a value, and not detect it at compile time. cc @RalfJung

Reply:
If you replace `()` with `u8` elsewhere, you can use `()` here which should continue to allow detection while being safer.

Reply:
I mean, yes, creating a `!` is insta-UB... is there unsafe code here that could accidentally do that?

Reply:
Oh there is compiler magic going on... I don't really know what to meaningfully respond here, @eddyb. Is there a question?

Reply:
Not really, I wrote that first comment before fully understanding the change.